bulletroboy/exoforce/exoforce_hardware.py

```diff
@@ -0,0 +1,266 @@
+import numpy as np
+from pyquaternion import Quaternion
+
+from .exoforce import ExoForce
+from ..utils.utils import Topics, Services
+
+from roboy_simulation_msgs.msg import Collision
+from roboy_middleware_msgs.msg import MotorCommand, MotorState
+from roboy_middleware_msgs.srv import ControlMode
+from geometry_msgs.msg import PoseStamped
+
+from rclpy.callback_groups import ReentrantCallbackGroup
+
+FAST_PWM_TENSION = 300
+MIN_PWM_TENSION = 150
+YES_OPTIONS = ["Yes","yes","y",""]
+
+POS_CONTROL_THRESHOLD = 0.1
+
+class ExoforceHW(ExoForce):
+	def __init__(self, cage_conf):
+		"""
+		Args:
+			cage_conf (CageConfiguration): Cage configuration defined in the configuration file.
+		
+		"""
+		super().__init__(cage_conf, "exoforce", POS_CONTROL_THRESHOLD)
+		self.callback_group = ReentrantCallbackGroup()
+
+		self.create_subscription(Collision, Topics.MAPPED_COLLISIONS, self.collision_listener, 1)
+		self.create_subscription(Collision, 'roboy/simulation/roboy/collision_hw', self.collision_listener, 1)
+		# self.create_subscription(PoseStamped, Topics.OP_EF_POSES, self.ef_pos_listener, 1)
+		self.create_subscription(PoseStamped, Topics.VR_HEADSET_POSES, self.operator_ef_pos_listener, 10, callback_group=self.callback_group)
+		
+		self.roboy_plexus_ready = False
+		choice = input("Connect to roboy plexus? (Yes/no): ")
+		if choice in YES_OPTIONS:
+			self.create_subscription(MotorState, Topics.MOTOR_STATE, self.motor_state_listener, 1, callback_group=self.callback_group)
+			self.motor_command_publisher = self.create_publisher(MotorCommand, Topics.MOTOR_COMMAND, 1)
+
+			self.control_mode_client = self.create_client(ControlMode, Services.CONTROL_MODE)
+			while not self.control_mode_client.wait_for_service(timeout_sec=1.0):
+				self.get_logger().info('roboy_plexus not available, waiting again...')
+			self.get_logger().info('Succesfull connection to roboy plexus!')
+			self.roboy_plexus_ready = True
+
+			self.control_mode_req = ControlMode.Request()
+			self.motor_command_msg = MotorCommand()
+
+			self.set_control_mode(3)
+
+			self.init_tendons()
+
+		choice = input("\nCalibrate motor positions? (Yes/no): ")
+		if choice in YES_OPTIONS:
+			self.calibration_timer = self.create_timer(2, self.calibration)
+		else:
+			self.start_exoforce()
+
+		self.applying_force = False
+		self.applying_force_timer = self.create_timer(1, self.stop_applying_force)
+		self.applying_force_timer.cancel()
+
+	def stop_applying_force(self):
+		self.applying_force = False
+
+	def start_exoforce(self):
+		self.create_timer(0.5, self.update)
+		choice = input("\nInitialize position control? (Yes/no): ")
+		if choice in YES_OPTIONS:
+			self.init_pos_control()
+
+	def init_tendons(self):
```

Comment:
what if i want to initialize only left hand tendons?

Reply:
We did not consider this functionality and would be a waste of time to implement right now, because of https://github.com/Roboy/bulletroboy/pull/25#issuecomment-760128816
If needed it can be added when the exoforce node is integrated with the force control node.